File: Leetcode/@Contest/Biweekly/Biweekly Contest 130/c.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 1005;
const int INF = 0x3f3f3f3f;

class Solution {
public:
    int dp[N];

    int dfs(int i, string &s) {
        if (i == s.size()) return 0;
        if (dp[i] != -1) return dp[i];
        int ans = INF, cnt[26] = {0}; 
        int mx, mn;
        for (int j = i; j < s.size(); j++) {
            cnt[s[j] - 'a'] += 1;
            mx = -INF, mn = INF;
            for (int k = 0; k < 26; k++) {
                if (cnt[k]) {
                    mx = max(mx, cnt[k]);
                    mn = min(mn, cnt[k]);
                }
            }
            if (mx == mn)
                ans = min(ans, 1 + dfs(j + 1, s));
        }
        return dp[i] = ans;
    }

    int minimumSubstringsInPartition(string s) {
        memset(dp, -1, sizeof(dp));
        return dfs(0, s);
    }
};

File: Leetcode/@Contest/Biweekly/Biweekly Contest 130/c.py
import math
from typing import *
from collections import *
from functools import *
from heapq import *
from bisect import *
from itertools import *

class Solution:
    def minimumSubstringsInPartition(self, s: str) -> int:
        n = len(s)

        @cache
        def dfs(i): # 從 i 開始劃分
            if i == n:
                return 0
            res = float("inf")
            cnt = [0] * 26
            for j in range(i, n):
                cnt[ord(s[j]) - ord("a")] += 1
                mx, mn = -float("inf"), float("inf")
                for k in range(26):
                    if cnt[k]:
                        mx = max(mx, cnt[k])
                        mn = min(mn, cnt[k])
                if mx == mn:
                    res = min(res, 1 + dfs(j + 1))
            return res
        return dfs(0)
    
sol = Solution()
print(sol.minimumSubstringsInPartition("aabb")) # 1
print(sol.minimumSubstringsInPartition("fabccddg")) # 3
print(sol.minimumSubstringsInPartition("abababaccddb")) # 2